torch/csrc/autograd/python_legacy_variable.cpp

```diff
@@ -42,27 +42,33 @@ static PyObject *THPVariable_pynew(PyTypeObject* type, PyObject *args, PyObject
     throw TypeError("_grad_fn has to be a Function object or None, but got %s",
         Py_TYPE(grad_fn)->tp_name);
   }
-  Tensor tensor;
+  Variable var;
   if (!data || data == Py_None) {
     // For legacy serialization code, create an empty tensor. This is also used
     // by nn.Parameter() with no arguments.
     auto scalar_type = torch::tensors::get_default_scalar_type();
-    auto var = at::empty({0}, torch::tensors::get_default_tensor_type().options(scalar_type));
-    tensor = static_cast<Variable&>(var).tensor_data();
+    var = at::empty({0}, torch::tensors::get_default_tensor_type().options(scalar_type));
   } else if (THPVariable_Check(data)) {
-    tensor = ((THPVariable*)data)->cdata.tensor_data();
+    var = ((THPVariable*)data)->cdata.variable_data();
   } else {
     throw torch::TypeError("Variable data has to be a tensor, but got %s",
         Py_TYPE(data)->tp_name);
   }
+  // We set `tensor`'s `allow_tensor_metadata_change` to true here, because we want to
+  // allow the following use case for backward compatibility:
+  //
+  // ```python
+  // var = Variable(torch.randn(2, 3))
+  // var.resize_(4, 5)
+  // ```
+  var.unsafeGetTensorImpl()->set_allow_tensor_metadata_change(true);
```

Comment:
@gchanan Thanks for the reminder - I updated the PR and put all the code samples in the code.

The difference between `variable_data` and `detach` is that `detach` shares version counter, while `variable_data` sets the new version counter to 0. Same as `variable_data`, `detach` also sets `allow_tensor_metadata_change` to false, and we have to manually set `allow_tensor_metadata_change` to true to allow for the mentioned use case.

The original semantics `var = make_variable(std::move(tensor), std::move(edge))` expects a new version counter to be created for `var`, so I think we should use `variable_data` instead of `detach` for this case.

Reply:
is this message out of date now?  (because you are using `detach` above).  I'm unclear if this is intentional or not (it's probably fine, although would be BC-breaking).  How about a comment about why you use or the other in the actual code, so we don't these two parallel tracks of conversation that make it difficult to figure out what the latest information is.